### PR TITLE
feat: Huobi provider to WS

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Modified Huobi provider to use [WebSocket](https://huobiapi.github.io/docs/spot/v1/en/#websocket-market-data) connection instead of `http.client`
- Modified unit tests for checking the map

ref: #509

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:159]:
```suggestion
// pong return a heartbeat message when a "ping" is received
```

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:207]:
```suggestion
		return TickerPrice{}, fmt.Errorf("failed to get ticker price for %s", cp.String())
```

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:215]:
```suggestion
// All data returned from the websocket Market APIs is compressed
// with GZIP, so it needs to be unzipped.
```

**Review by adamewozniak — COMMENTED** [review]
just a couple nits, plus : 
1. Please add a line to `CHANGELOG.md` 
2. Why do we not need reconnection logic like we have with binance & okx?

**Comment by RafilxTenfen** [comment]
> * Please add a line to `CHANGELOG.md`

My bad forgot it again :/

**Comment by adamewozniak** [comment]
> @adamewozniak
> 
> > 2. Why do we not need reconnection logic like we have with binance & okx?
> 
> * Huobi handles it differently, it maintains the connection alive without any time constraint, while we still respond to the heartbeat message [ping x pong](https://github.com/umee-network/umee/pull/569/files#diff-88966ac3f8eb0c4b434ceef0aac6226c51a21863af1602e40cc085ddaf14e0ecR166)
> 
> [huobi doc](https://huobiapi.github.io/docs/spot/v1/en/#introduction-10)

Per the huobi docs : 
> After the server sent two consecutive heartbeat messages without receiving at least one matching "pong" response from a client, then right before server sends the next "ping" heartbeat, the server will be disconnected with the client server.

So then, if for some reason the price feeder didn't send the pong response to huobi within that window, the user would have to restart the program to reconnect? In the case of an intermittent outage, or the price feeder is unable to send that response for whatever reason, we wouldn't want to require validators to restart the price feeder in order to reconnect huobi

**Comment by RafilxTenfen** [comment]
> So then, if for some reason the price feeder didn't send the pong response to huobi within that window, the user would have to restart the program to reconnect? In the case of an intermittent outage, or the price feeder is unable to send that response for whatever reason, we wouldn't want to require validators to restart the price feeder in order to reconnect huobi 

- ohh yeah, you are correct, thanks for clarifying 
- I will think of something to handle that 
- until there I will set the PR as draft

**Comment by RafilxTenfen** [comment]
@adamewozniak 

I did 2 things to try to reconnect, one of them is as usual, try to reconnect at each 2 min if the ticker is not reset by a ping message, the other one became a litter [weird](https://github.com/umee-network/umee/pull/569/files#diff-88966ac3f8eb0c4b434ceef0aac6226c51a21863af1602e40cc085ddaf14e0ecR114) 
- if a message can not be read it generates an error in this case we try to ping the provider to check the connection 
   - if the ping returns an error it means that the connection is closed so I try to reconnect there too

Any thoughts around that?

**Comment by adamewozniak** [comment]
> @adamewozniak
> 
> I did 2 things to try to reconnect, one of them is as usual, try to reconnect at each 2 min if the ticker is not reset by a ping message, the other one became a little [weird](https://github.com/umee-network/umee/pull/569/files#diff-88966ac3f8eb0c4b434ceef0aac6226c51a21863af1602e40cc085ddaf14e0ecR114)
> 
> * if a message can not be read it generates an error in this case we try to ping the provider to check the connection
>   
>   * if the ping returns an error it means that the connection is closed so I try to reconnect there too
> 
> Any thoughts around that?

I think it makes sense to handle the case this way - I'm assuming the err coming out of `p.wsClient.ReadMessage()` is rare / would only happen if the provider was down, yes?

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:120]:
```suggestion
						p.logger.Err(err).Msg("error reconnecting to the huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:134]:
```suggestion
				p.logger.Err(err).Msg("error reconnecting to the huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:217]:
```suggestion
		return fmt.Errorf("error reconnecting to huobi websocket: %w", err)
```

**Review by adamewozniak — APPROVED** [review]
just some nits, but LGTM 🚀 might want a review from @alexanderbez before merging

**Comment by RafilxTenfen** [comment]
> > @adamewozniak
> > I did 2 things to try to reconnect, one of them is as usual, try to reconnect at each 2 min if the ticker is not reset by a ping message, the other one became a little [weird](https://github.com/umee-network/umee/pull/569/files#diff-88966ac3f8eb0c4b434ceef0aac6226c51a21863af1602e40cc085ddaf14e0ecR114)
> > 
> > * if a message can not be read it generates an error in this case we try to ping the provider to check the connection
> >   
> >   * if the ping returns an error it means that the connection is closed so I try to reconnect there too
> > 
> > Any thoughts around that?
> 
> I think it makes sense to handle the case this way - I'm assuming the err coming out of `p.wsClient.ReadMessage()` is rare / would only happen if the provider was down, yes?

Usually, yes 
I had it running for some time and only generated this error when the connection was closed

File: price-feeder/oracle/provider/binance.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/binance.go:90]:
```suggestion
// GetTickerPrices returns the tickerPrices based on the provided pairs.
```

Comment (alexanderbez) [reply price-feeder/oracle/provider/binance.go:90]: good catch :P

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:33]:
```suggestion
	// REF: https://huobiapi.github.io/docs/spot/v1/en/#market-ticker
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:41]: same as the previous PR -- this doesn't really have a need to be a part of the `HuobiProvider` type.

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:45]:
```suggestion
	// the tick object for a given ticker/symbol.
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:52]:
```suggestion
	// and the last traded price for a given ticker/symbol.
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:87]: why not have this be part of the struct literal creation on line 76?

Comment (alexanderbez) [reply price-feeder/oracle/provider/huobi.go:87]: You can update the other providers too :)

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:111]: Can we rename this to `defaultReadNewWSMessage`?

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:112]: What does this comment mean?

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:116]:
```suggestion
				p.logger.Err(err).Msg(" failed to read message from Huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:118]:
```suggestion
					p.logger.Err(err).Msg("failed to send ping")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:120]:
```suggestion
						p.logger.Err(err).Msg("error reconnecting to the Huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:134]:
```suggestion
				p.logger.Err(err).Msg("error reconnecting to the Huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:140]:
```suggestion
// messageReceived handles the received data from the Huobi websocket.
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:150]:
```suggestion
		p.logger.Err(err).Msg("failed to decompress Huobi gziped message")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:142]:
```suggestion
// GZIP so they need to be decompressed.
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:154]: let's put `[]byte("ping")` in a global private `var`?

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:162]:
```suggestion
		p.logger.Err(err).Msg("failed to unmarshal message from Huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:251]:
```suggestion
func decompressGzip(bz []byte) ([]byte, error) {
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:248]:
```suggestion
// decompressGzip uncompress gzip compressed messages
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Looks great @RafilxTenfen. Just left a bit of minor nitpick feedback ;-)

File: price-feeder/oracle/provider/huobi.go
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/huobi.go]: done at https://github.com/umee-network/umee/pull/569/commits/576f19d2c689b558508636c8419600b5c3c75aaf

File: price-feeder/oracle/provider/huobi.go
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/huobi.go]: done at https://github.com/umee-network/umee/pull/569/commits/9c7f0c6af5cf96bbafdb3373fe52aca44e51b2b7

File: price-feeder/oracle/provider/huobi.go
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/huobi.go]: done at https://github.com/umee-network/umee/pull/569/commits/d90973ee0251cf9d4e4caf1b6384192d10084a36

File: price-feeder/oracle/provider/huobi.go
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/huobi.go]: It was kind of to give a reason to why have the `time.After` 
Removed at https://github.com/umee-network/umee/pull/569/commits/9269b0c910897b95d3ce3a279368d7e45b31339a

File: price-feeder/oracle/provider/huobi.go
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/huobi.go]: done at https://github.com/umee-network/umee/pull/569/commits/e9099221b3884748aed6f8862b70836b3a2270e7

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:115]:
```suggestion
				p.logger.Err(err).Msg("failed to read message from Huobi provider")
```

File: price-feeder/oracle/provider/huobi.go
Comment (alexanderbez) [inline price-feeder/oracle/provider/huobi.go:114]:
```suggestion
				// If some error occurs, check if connection is alive
				// and continue to try to read the next message.
```

**Review by alexanderbez — APPROVED** [review]
Nice work @RafilxTenfen 👏

**Comment by alexanderbez** [comment]
> @Mergifyio backport release/price-feeder/v0.1.x

@RafilxTenfen you only need to add this comment _iff_ the PR was already _merged_. While it's still open, you can just simply add a `backport` label to the PR (I added it). 👍

**Comment by RafilxTenfen** [comment]
> > @Mergifyio backport release/price-feeder/v0.1.x
> 
> @RafilxTenfen you only need to add this comment _iff_ the PR was already _merged_. While it's still open, you can just simply add a `backport` label to the PR (I added it). +1

Thanks!

**Pull request merged**
